crates/runtime/src/tools/builtin/table_schema.rs

```diff
@@ -44,6 +62,71 @@ impl TableSchemaTool {
             description,
         }
     }
+
+    pub async fn get_schema(
+        &self,
+        df: Arc<DataFusion>,
+        req: &TableSchemaToolParams,
+    ) -> Result<Value, Box<dyn std::error::Error + Send + Sync>> {
+        let span = tracing::span!(target: "task_history", tracing::Level::INFO, "tool_use::table_schema", tool = self.name(), input = serde_json::to_string(&req).boxed()?);
+
+        let mut table_schemas: Vec<Value> = Vec::with_capacity(req.tables.len());
+        for t in &req.tables {
+            let schema = df
+                .get_arrow_schema(t)
+                .instrument(span.clone())
+                .await
+                .boxed()?;
+
+            let schema_value = serde_json::value::to_value(schema).boxed()?;
+
+            let table_schema = serde_json::json!({
```

Comment:
why not use an intermediary struct, like `TableSchemaResponse`? This would avoid needing to JSON serialise every element, when we JSON serialise again on L91 anyway. I'm mainly concerned about performance of many JSON serialises here.

Reply:
It's not quite serializing it to a JSON string, but creating a `serde_json::Value` struct instead. The stringification only happens on line 91